File: bundles/io.github.linkedfactory.service/src/main/scala/io/github/linkedfactory/service/KvinManager.scala
package io.github.linkedfactory.service

import io.github.linkedfactory.kvin.Kvin
import io.github.linkedfactory.kvin.influxdb.KvinInfluxDb
import io.github.linkedfactory.kvin.leveldb.KvinLevelDb
import net.enilink.komma.core.{IReference, URIs}
import net.enilink.komma.em.concepts.IResource
import net.enilink.platform.lift.util.Globals
import org.eclipse.core.runtime.Platform
import org.osgi.framework.FrameworkUtil

import java.io.File
import java.nio.file.Files
import scala.util.{Failure, Success, Try}

object KvinManager {
  val bundleContext = Option(FrameworkUtil.getBundle(getClass)).map(_.getBundleContext).getOrElse(null)
  val instanceLoc = if (bundleContext != null) Platform.getInstanceLocation else null
  val cfgURI = URIs.createURI("plugin://de.fraunhofer.iwu.linkedfactory.service/data/")
  private var kvin: Option[Kvin] = None

  def getKvin(): Option[Kvin] = kvin

  {
    Globals.withPluginConfig { pcModel => {
      val cfg = pcModel.getManager.find(cfgURI.appendLocalPart("store")).asInstanceOf[IResource]
      kvin = cfg.getSingle(cfgURI.appendLocalPart("type")) match {
        case s: String if (s == "InfluxDB") =>
          val endpoint = cfg.getSingle(cfgURI.appendLocalPart("endpoint")) match {
            case r: IReference if r.getURI != null => r.getURI
            case s: String => URIs.createURI(s)
            case _ => URIs.createURI("http://localhost:8086/")
          }
          val db = cfg.getSingle(cfgURI.appendLocalPart("database")) match {
            case s: String => s
            case _ => "LinkedFactory"
          }
          Try(new KvinInfluxDb(endpoint, db)) match {
            case Failure(throwable) =>
              System.err.println(s"Value store: FAILURE for InfluxDB @ $endpoint: ${throwable.getMessage}")
              None
            case Success(success) =>
              println(s"Value store: using InfluxDB @ $endpoint")
              Some(success)
          }
        //case s: String if (s == "KVIN") =>
        case _ =>
          val dirName = cfg.getSingle(cfgURI.appendLocalPart("dirName")) match {
            case s: String => s
            case _ => "linkedfactory-valuestore"
          }
          val valueStorePath = new File(dirName) match {
            case dir if dir.isAbsolute() => dir
            case _ if instanceLoc.isSet => new File(new File(instanceLoc.getURL.toURI), dirName)
            case _ => new File(Files.createTempDirectory("lf").toFile, dirName)
          }
          Try {
            new KvinLevelDb(valueStorePath)
          } match {
            case Failure(throwable) =>
              System.err.println(s"Value store: FAILURE for LF w/ path=$valueStorePath: ${throwable.getMessage}")
              None
            case Success(success) =>
              println(s"Value store: using LF w/ path=$valueStorePath")
              Some(success)
          }
      }
    }
    }
  }
}

File: bundles/io.github.linkedfactory.service/src/main/scala/io/github/linkedfactory/service/KvinManager.java
package io.github.linkedfactory.service;

import io.github.linkedfactory.core.kvin.Kvin;
import io.github.linkedfactory.service.config.IKvinFactory;
import io.github.linkedfactory.service.config.KvinLevelDbFactory;
import net.enilink.komma.core.URI;
import net.enilink.komma.core.URIs;
import net.enilink.komma.em.concepts.IClass;
import net.enilink.komma.em.concepts.IResource;
import net.enilink.platform.core.PluginConfigModel;
import org.osgi.framework.ServiceRegistration;
import org.osgi.service.component.ComponentContext;
import org.osgi.service.component.annotations.Activate;
import org.osgi.service.component.annotations.Component;
import org.osgi.service.component.annotations.Deactivate;
import org.osgi.service.component.annotations.Reference;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Hashtable;

@Component
public class KvinManager {

    static final Logger log = LoggerFactory.getLogger(KvinManager.class);

    static final URI cfgUri = URIs.createURI("plugin://io.github.linkedfactory.service/data/");
    PluginConfigModel configModel;
    ServiceRegistration<Kvin> kvinServiceRegistration;
    Kvin kvin;

    @Reference
    void setConfigModel(PluginConfigModel configModel) {
        this.configModel = configModel;
    }

    @Activate
    void activate(ComponentContext ctx) {
        try {
            configModel.begin();
            Object cfg = configModel.getManager().find(cfgUri, IResource.class)
                    .getSingle(cfgUri.appendLocalPart("store"));

            if (! (cfg instanceof IResource)) {
                log.error("Kvin store is not properly configured in: {}", cfgUri);
                return;
            }

            IResource cfgResource = (IResource) cfg;
            IKvinFactory factory;
            if (!(cfg instanceof IKvinFactory)) {
                if (cfgResource.getRdfTypes().isEmpty()) {
                    // fallback to KvinLevelDb
                    log.info("Using default KvinLevelDb as type is not specified in config: {}", cfgUri);
                    cfgResource.getRdfTypes().add(configModel.getManager()
                            .find(URIs.createURI(KvinLevelDbFactory.TYPE), IClass.class));
                    factory = cfgResource.as(IKvinFactory.class);
                } else {
                    log.error("Invalid Kvin configurations with types: {}", cfgResource.getRdfTypes());
                    return;
                }
            } else {
                factory = (IKvinFactory) cfg;
            }

            try {
                kvin = factory.create();
                kvinServiceRegistration = ctx.getBundleContext().registerService(Kvin.class, kvin, new Hashtable<>());
            } catch (Throwable throwable) {
                log.error("Failure while creating Kvin store", throwable);
            }
        } finally {
            configModel.end();
        }
    }

    @Deactivate
    void deactivate(ComponentContext ctx) {
        if (kvinServiceRegistration != null) {
            kvinServiceRegistration.unregister();
            kvinServiceRegistration = null;
        }
        if (kvin != null) {
            kvin.close();
            kvin = null;
        }
    }

}
